homeassistant/components/stream/hls.py

```diff
@@ -64,55 +78,184 @@ class HlsPlaylistView(StreamView):
     cors_allowed = True
 
     @staticmethod
-    def render_preamble(track):
-        """Render preamble."""
-        return [
-            "#EXT-X-VERSION:7",
-            f"#EXT-X-TARGETDURATION:{track.target_duration}",
-            '#EXT-X-MAP:URI="init.mp4"',
-        ]
+    def render(track, ll_hls):
+        """Render HLS playlist file."""
+        # NUM_PLAYLIST_SEGMENTS+1 because most recent is probably not yet complete
+        segments = list(track.get_segment())[-(NUM_PLAYLIST_SEGMENTS + 1) :]
 
-    @staticmethod
-    def render_playlist(track):
-        """Render playlist."""
-        segments = list(track.get_segment())[-NUM_PLAYLIST_SEGMENTS:]
+        # Create a copy of the last segment with the parts frozen for rendering playlist
+        segments[-1] = copy(segments[-1])
+        segments[-1].parts = segments[-1].parts.copy()
+
+        # Remove first segment if the last segment is actually complete
+        if segments[-1].duration:
+            segments = segments[-NUM_PLAYLIST_SEGMENTS:]
 
-        if not segments:
-            return []
+        target_duration = round(max(s.duration for s in segments if s.duration))
 
         playlist = [
+            "#EXTM3U",
+            "#EXT-X-VERSION:6",
+            "#EXT-X-INDEPENDENT-SEGMENTS",
+            '#EXT-X-MAP:URI="init.mp4"',
+            f"#EXT-X-TARGETDURATION:{target_duration}",
             f"#EXT-X-MEDIA-SEQUENCE:{segments[0].sequence}",
             f"#EXT-X-DISCONTINUITY-SEQUENCE:{segments[0].stream_id}",
+            "#EXT-X-PROGRAM-DATE-TIME:"
+            + segments[0].start_time.strftime("%Y-%m-%dT%H:%M:%S.%f")[:-3]
+            + "Z",
         ]
 
-        last_stream_id = segments[0].stream_id
-        for segment in segments:
-            if last_stream_id != segment.stream_id:
-                playlist.append("#EXT-X-DISCONTINUITY")
+        if ll_hls:
+            part_duration = float(
+                max(
+                    (p.duration for s in segments for p in s.parts),
+                    default=PART_DURATION,
+                )
+            )
             playlist.extend(
                 [
-                    f"#EXTINF:{float(segment.duration):.04f},",
-                    f"./segment/{segment.sequence}.m4s",
+                    f"#EXT-X-PART-INF:PART-TARGET={part_duration:.3f}",
+                    f"#EXT-X-SERVER-CONTROL:CAN-BLOCK-RELOAD=YES,PART-HOLD-BACK={2*part_duration:.3f}",
```

Comment:
From 4.4.3.8 on PART-HOLD-BACK:

> The value is a decimal-floating-point number of seconds that
>       indicates the server-recommended minimum distance from the end of
>       the Playlist at which clients should begin to play or to which
>       they should seek when playing in Low-Latency Mode.  Its value MUST
>       be at least twice the Part Target Duration.  Its value SHOULD be
>       at least three times the Part Target Duration.  If different
>       Renditions have different Part Target Durations then PART-HOLD-
>       BACK SHOULD be at least three times the maximum Part Target
>       Duration.

So we just use the minimum allowed value of 2 `PART_DURATIONS`. For EXT-START-TIME, I just added an extra `PART_DURATION` (for a total of 3 `PART_DURATIONS`) under the untested assumption that starting a little further from the live edge would help smoothness. We can consider tweaking this value.